appengine/src/test/java/com/google/googleinterns/gscribe/services/ExamServiceTest.java

```diff
@@ -0,0 +1,490 @@
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+public class ExamServiceTest {
+
+    private final static User user = new User("user", "access_token", "refresh_token", null);
+    private final static String spreadsheetId = "spreadsheet_id";
+    private final static String sheetName = "sheet_name";
+    private final static ExamRequest examRequest = new ExamRequest();
+    private final static ValueRange valueRangeForNumberOfQuestions = new ValueRange();
+    @Mock
+    SpreadsheetService spreadsheetService;
+    @InjectMocks
+    ExamServiceImpl examService;
+
+    @BeforeAll
+    private static void init() {
+        examRequest.setSheetName(sheetName);
+        examRequest.setSpreadsheetID(spreadsheetId);
+        List<List<Object>> valueRangeDataFoNumberOfQuestions = new ArrayList<>();
+        valueRangeDataFoNumberOfQuestions.add(Arrays.asList(new String[]{"Duration of exam:"}));
+        valueRangeDataFoNumberOfQuestions.add(Arrays.asList(new String[]{"Question Type"}));
+        valueRangeDataFoNumberOfQuestions.add(Arrays.asList(new String[]{"SUBJECTIVE"}));
+        valueRangeDataFoNumberOfQuestions.add(Arrays.asList(new String[]{"MCQ"}));
+        valueRangeForNumberOfQuestions.setValues(valueRangeDataFoNumberOfQuestions);
+    }
+
+
+    @Test
+    public void createNewExamTest() throws GeneralSecurityException, InvalidDatabaseDataException, InvalidRequestException, IOException, ExamFormatException {
+
+        List<List<Object>> valueRangeDataWhole = new ArrayList<>();
+        valueRangeDataWhole.add(Arrays.asList(new String[]{"Duration of exam:", "100"}));
+        valueRangeDataWhole.add(Arrays.asList(new String[]{"Question Type", "Question statement", "Option A", "Option B", "Option C", "Option D", "Points"}));
+        valueRangeDataWhole.add(Arrays.asList(new String[]{"SUBJECTIVE", "SubjectiveStatement1", "", "", "", "", "2"}));
+        valueRangeDataWhole.add(Arrays.asList(new String[]{"MCQ", "MCQStatement1", "OptionA", "OptionB", "OptionC", "OptionD", "3"}));
+        ValueRange valueRangeWhole = new ValueRange();
+        valueRangeWhole.setValues(valueRangeDataWhole);
+
+        when(spreadsheetService.parseSpreadsheetRequest(user, spreadsheetId, sheetName + "!A:A")).thenReturn(valueRangeForNumberOfQuestions);
+        when(spreadsheetService.parseSpreadsheetRequest(user, spreadsheetId, sheetName + "!A1:G4")).thenReturn(valueRangeWhole);
+
+        Exam exam = examService.getExam(examRequest, user);
+
+        assertEquals(exam.getExamMetadata().getSpreadsheetID(), spreadsheetId);
+        assertEquals(exam.getExamMetadata().getDuration(), 100);
+        assertEquals(exam.getExamMetadata().getUserID(), user.getId());
+        assertEquals(exam.getQuestions().getQuestionsList().size(), 2);
+        assertEquals(exam.getQuestions().getQuestionsList().get(0).getType(), QuestionType.SUBJECTIVE);
```

Comment:
Override equals and hashCode methods instead of comparing each and every field.

Reply:
Created equals() and hashCode() for all models and updated test accordingly.

Reply:
After updating equals and hashCode, you should not check assertEquals for each and every field individually.

Reply:
I updated the code accordingly but the change was not getting rendered here. It is showing in the files but not here in the files changed.